### ISSUE TITLE
More comments

### ISSUE THREAD

**Issue by dpo** [CiDAMO/AugLag.jl#17]
### Model

- [x] It's not a good idea to have abstract members in a struct: https://github.com/CiDAMO/AugLag.jl/blob/master/src/AugLagModel.jl#L16
- [x] check that initial mu is ≥ 0 and y has the correct size https://github.com/CiDAMO/AugLag.jl/blob/master/src/AugLagModel.jl#L22
- [x] the cost of a `jtprod()` is normally higher than that of adding two vectors, so [this](https://github.com/CiDAMO/AugLag.jl/blob/master/src/AugLagModel.jl#L47) should probably be written `jtprod(nlp.model, x, nlp.y - nlp.mu * nlp.cx)`
- [x] perhaps you could store `nlp.y - nlp.mu * nlp.cx` inside the model because it is recomputed several times per iteration

### Algorithm

- [x] the unconstrained case should probably be implemented as a trait:
```julia
_is_unconstrained(nlp::AbstractNLPModel) = unconstrained(nlp) ? Val{true}() : Val{false}()

function al(nlp :: AbstractNLPModel, args...; kwargs...) = al(_is_unconstrained(nlp), nlp, args...; kwargs...)

al(Val{true}, nlp :: AbstractNLPModel, args...; kwargs...) = tron(nlp, args...; kwargs...)

function al(Val{false}, nlp :: AbstractNLPModel, args...; kwargs...)
  # ... your implementation for the constrained case
end
```
- [ ] similarly, you could use traits to avoid creating a `SlackModel` inside the algorithm
- [x] the user should be able to specify a logger for the subproblem: https://github.com/CiDAMO/AugLag.jl/blob/master/src/al.jl#L81
- [x] you don't want to be converting `lvar` and `uvar` at each iteration: https://github.com/CiDAMO/AugLag.jl/blob/master/src/al.jl#L99
- [x] note that TRON will have computed `norm(cx)` as part of the most recent objective evaluation: https://github.com/CiDAMO/AugLag.jl/blob/master/src/al.jl#L86
- [x] I don't think this computation occurs in type `T`: https://github.com/CiDAMO/AugLag.jl/blob/master/src/al.jl#L90

**Comment by abelsiqueira** [comment]
@dpo:
- 1st points: do you mean `AbstractNLPModel` as well, or just `AbstractVector`s?
- 5th and 6th points: Something like
```julia
function al(nlp)
    if unconstrained(nlp) || bound_constrained(nlp)
        al(Val(:tron), nlp)
    elseif inequality_constrained(nlp)
        al(Val(:ineq), nlp)
    else
        al(Val(:al), nlp)
    end
end

al(Val{:tron}, nlp) = tron(nlp)
function al(Val{:ineq}, nlp)
    snlp = SlackModel(nlp)
    output = al(Val{:ineq}, nlp)
    # Convert output back
    return output
end
function al(Val{:al}, nlp)
    # ...
end
```

**Comment by dpo** [comment]
> 1st points: do you mean AbstractNLPModel as well, or just AbstractVectors?

All abstract fields I think. I would be tempted to write something like
```julia
mutable struct AugLagModel{M,F,V} <: AbstractNLPModel where {M <: AbstractNLPModel, F <: AbstractFloat, V <: AbstractVector}
  meta :: NLPModelMeta
  counters :: Counters
  model :: M
  y :: V
  mu :: F
  x :: V # save last iteration of subsolver
  cx :: V # save last constraint value of subsolver
end
```
It's tricky however. See https://discourse.julialang.org/t/struct-with-abstract-type-field/8743

> 5th and 6th points: Something like

👍

**Comment by abelsiqueira** [comment]
Closes after #26
